internal/dnfjson/cache.go

```diff
@@ -1,46 +1,187 @@
 package dnfjson
 
 import (
+	"fmt"
 	"io/fs"
 	"os"
 	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/gobwas/glob"
 )
 
 type rpmCache struct {
 	// root path for the cache
 	root string
 
+	// individual repository cache data
+	repoElements map[string]pathInfo
+
+	// list of known repository IDs, sorted by mtime
+	repoRecency []string
+
+	// total cache size
+	size uint64
+
 	// max cache size
 	maxSize uint64
 }
 
 func newRPMCache(path string, maxSize uint64) *rpmCache {
-	return &rpmCache{
-		root:    path,
-		maxSize: maxSize,
+	r := &rpmCache{
+		root:         path,
+		repoElements: make(map[string]pathInfo),
+		size:         0,
+		maxSize:      maxSize,
 	}
+	// collect existing cache paths and timestamps
+	r.updateInfo()
+	return r
 }
 
-func (r *rpmCache) size() (uint64, error) {
-	var size uint64
-	sizer := func(path string, info fs.FileInfo, err error) error {
+func (r *rpmCache) clean() error {
+	curSize, err := dirSize(r.root)
+	if err != nil {
+		return err
+	}
+	if curSize > r.maxSize {
+		return os.RemoveAll(r.root)
+	}
+	return nil
+}
+
+// updateInfo updates the repoPaths and repoRecency fields of the rpmCache.
+func (r *rpmCache) updateInfo() {
+	repos := make(map[string]pathInfo)
+	repoIDs := make([]string, 0)
```

Comment:
It would be more efficient to first get `cacheEntries` and use its `len()` when calling `make()` to create `repos` and `repoIDs`. Currently, there is unnecessary resizing happening when adding elements to the slice and map.

Reply:
Considered doing something "smart" like allocating `len(cacheEntries)/3` (which is the average number of entries per cache), but I just went with a plain `len(cacheEntries)` instead.

Reply:
Yeah, I meant the capacity. I must admit that I missed the fact that there would be more than one cache entries per repo (it was not that obvious from the code at the first sight), but after your comment and connecting the dots with the unit test, I see it now.